### PR TITLE
Add ability to support "required" flag in module dependencies of module.xml

### PR DESCRIPTION
**Pull request by ia-keaton**
This PR introduces a new flag to the `ModuleSettings` class called `requiredModuleDependencies`. This flag is intended as a substitute for the existing `moduleDependencies`, not a replacement. Current module builds should be unaffected.

In order to utilize this flag, `requiredIgnitionVersion` will need to also be set to at least 8.3.0. To simplify the usage of this new property, this PR also adds a Kotlin DSL to parse user input into a new `RequiredModuleDependency` class. Usage of this DSL is in the following form:
```
requiredModuleDependencies = setOf(
     moduleId("com.inductiveautomation.vision") {
         scope = "GCD"
         required = true
     }
)
```

This PR also includes unit tests for the `WriteModuleXml` class which previously had none. These tests are for usage of both the new flag as well as the original.

### CODE REVIEW

File: gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/extension/RequiredModuleDependency.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package io.ia.sdk.gradle.modl.extension
+
+import java.io.Serializable
+
+class RequiredModuleDependency(
```
Comment (brianeray) [inline gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/extension/RequiredModuleDependency.kt:5]: Realized after seeing all the code together that `RequiredModuleDependency` and `requiredModuleDependencies` are confusing, since `required` can be--and is by default--`false`.

A straightforward change would be just renaming to something like `ModuleDependencySpecification` and `moduleDependencySpecifications` or similar, implying it's a more structured flavor of `moduleDependencies` and its `MapProperty<String,String>` type.

I'm going to see if a trick is possible to smooth this deprecation further. Will take me a bit but might pay off.

Comment (brianeray) [reply gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/extension/RequiredModuleDependency.kt:5]: Before I look for that trick, my branch is now synced with yours, plus this [commit at the head](https://github.com/brianeray/ignition-module-tools/commit/8e9ce856c493aa2e2b37eda60514d5fecfb4f6b7) with the renaming above.

`gradlew check` is completely clean on that refactor.

File: gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/extension/ModuleSettings.kt [hunk 2]
```diff
@@ -83,6 +84,43 @@ open class ModuleSettings @javax.inject.Inject constructor(objects: ObjectFactor
...
+     *   moduleDependencySpecs = setOf(
+     *      moduleId("com.inductiveautomation.vision") {
+     *          scope = "GCD"
+     *          required = true
+     *      }
+     *   )
+     */
+    val moduleDependencySpecs: SetProperty<ModuleDependencySpec> = objects.setProperty(
+        ModuleDependencySpec::class.java
+    ).convention(
```
[the remark this answers is not in the document]
Comment (ia-keaton) [reply gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/extension/ModuleSettings.kt:112]: Yeah the convention worked for kotlin builds (it would default to getting its data from the legacy fields correctly and the additional conditional wasn't necessary in WriteModuleXML), but for some reason it didn't work with Groovy builds. Hence the additional conditional. I agree with taking it out in this case.

**Comment by brianeray** [comment]
I figured out the `compileTestKotlin` warnings, here in [6a5982a](https://github.com/brianeray/ignition-module-tools/commit/6a5982a04369fec3d12f52d74fd6540a19a068cf).

Pretty sure a similar build script bug bit me a long, long time ago.

Now onto a "final" PR pass.

File: gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt [hunk 3]
```diff
@@ -151,10 +164,20 @@ open class WriteModuleXml @Inject constructor(_objects: ObjectFactory) : Default
                     }
                 }
```
Comment (brianeray) [inline gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt:166]: We might want to trap and fail on `moduleDependencies.isPresent && moduleDependencySpecs.isPresent`  to shunt module devs in one direction or another. That'd be a hard guardrail.

Or we could allow both, and turn the `else if` below into a regular `if`. Wild west stuff, but still reasonable.

Or we could leave this conditional as is until somebody tries to use both and their `moduleDependencySpecs` are silently ignored, and then fix it the way they think it should work.

File: gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt [hunk 4]
```diff
@@ -188,6 +211,13 @@ open class WriteModuleXml @Inject constructor(_objects: ObjectFactory) : Default
         return modules.toString(PrintOptions(pretty = true, singleLineTextElements = true, useSelfClosingTags = false))
     }
 
+    private fun usemoduleDependencySpecs(): Boolean {
+        if (!requiredIgnitionVersion.isPresent) return false
+
+        val version = requiredIgnitionVersion.get().split(".")
+        return version.size >= 2 && version[0].toInt() >= 8 && version[1].toInt() >= 3
```
Comment (brianeray) [inline gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt:218]: I haven't tried this yet but there is a bug in the expression. It would not do what we want on, for example, `9.0.0`, `9.0`, or `9`.

```suggestion
        val version = requiredIgnitionVersion.get().split(".").map { it.toInt() }
        if (version[0] >= 8) { return true }
        if (version[0] == 8 && version[1] >= 3) { return true }
        return false
```

**Review by brianeray — CHANGES_REQUESTED** [review]

File: gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt (hunk 3 above)
Comment (ia-keaton) [reply gradle-module-plugin/src/main/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXml.kt:166]: Hmm alternatively I could just flip the conditionals so if both are present we use the newer one

**Review by brianeray — APPROVED** [review]
